### PR TITLE
[api][server] Fix comments for CodeChecker cmd results

### LINKED ISSUE #3005: Comments are not displayed in the Cmd results
**Describe the bug**
I want to export the results from Codechecker Server into JSON format. For doing that, I am using  `cmd results` command as:
```
CodeChecker cmd results smatch -o json 2>/dev/null | python -m json.tool > smatch_out.json
```

The output file( `smatch_out.json`) does not properly export the comment on bugs in Reports stored in `Codechecker server`. Instead of displaying the comment, only an empty string(`""`) is displayed.

Kindly provide pointers to solve this issue

### PR DESCRIPTION
**Pull request by jay24rajput**
## Drawback with the Previous System
The command `CodeChecker cmd results` was not able to fetch all the comments appropriately from the `CodeChecker server`. In many runs, the comments were displayed as `""`, thus making it impossible to use Codechecker as an exporter.

## Functionality Added
Comments are fetched and formatted appropriately. The comments are fetched with the command
```sh
CodeChecker cmd results <run_name> -o json --details
```
- [x] Implementation 
- [x] Tests

Fixes #3005

### CODE REVIEW

File: web/server/codechecker_server/api/report_server.py [hunk 1]
```diff
@@ -623,11 +623,26 @@ def get_report_details(session, report_ids):
...
+    comment_data_list = defaultdict(list)
+    comment_query = session.query(Comment, Report.id, File.filepath)\
+        .outerjoin(Report, Report.bug_id == Comment.bug_hash) \
+        .outerjoin(File,
+                   File.id == Report.file_id)
+
+    for data, report_id, file_path in comment_query:
+        report_id = report_id
+        comment_data = comment_data_db_to_api(data)
+        comment_data.filePath = file_path
```
Comment (csordasmarton) [inline web/server/codechecker_server/api/report_server.py:636]: Why do you try to set a field why doesn't even exists? The CommentData struct doesn't have a filePath field:
https://github.com/Ericsson/codechecker/blob/47cf966edcafaaaf1047298d8fe0b03f0f3ca82a/web/api/report_server.thrift#L303-L309

Comments are connected to Reports through the Report hash (bug id) field, and every Report points to a file path. So I think there is no need to query the file path and use it here.

File: web/server/codechecker_server/api/report_server.py [hunk 2]
```diff
@@ -623,11 +623,26 @@ def get_report_details(session, report_ids):
         extended_data.filePath = file_path
         extended_data_list[report_id].append(extended_data)
 
+    # Get Comments for report data
+    comment_data_list = defaultdict(list)
+    comment_query = session.query(Comment, Report.id, File.filepath)\
```
Comment (csordasmarton) [inline web/server/codechecker_server/api/report_server.py:628]: We need to filter the comments by the `report_ids` function parameter:
```py
.filter(Report.id.in_(report_ids)) \
```
This way we do not query everything from the database which is not used.

File: web/server/codechecker_server/api/report_server.py [hunk 3]
```diff
@@ -623,11 +623,26 @@ def get_report_details(session, report_ids):
...
+                   File.id == Report.file_id)
+
+    for data, report_id, file_path in comment_query:
+        report_id = report_id
+        comment_data = comment_data_db_to_api(data)
+        comment_data.filePath = file_path
+        comment_data_list[report_id].append(comment_data)
+
     for report_id in report_ids:
+        # print(comment_data_list[report_id])
```
Comment (csordasmarton) [inline web/server/codechecker_server/api/report_server.py:640]:
```suggestion
```

File: web/server/codechecker_server/api/report_server.py [hunk 4]
```diff
@@ -662,6 +677,16 @@ def extended_data_db_to_api(erd):
         fileId=erd.file_id)
 
 
+def comment_data_db_to_api(comm):
+    return ttypes.CommentData(
+        id=comm.id,
+        author=comm.author,
+        message=comm.message.decode("utf-8").replace("\\", ""),
```
Comment (csordasmarton) [inline web/server/codechecker_server/api/report_server.py:684]: For system comments we should post process the message. See:
https://github.com/Ericsson/codechecker/blob/47cf966edcafaaaf1047298d8fe0b03f0f3ca82a/web/server/codechecker_server/api/report_server.py#L1610-L1630
We can move this logic to a separate helper function and use it here and in the original place.

**Review by csordasmarton — CHANGES_REQUESTED** [review]
Please also add a new test cases here: 
https://github.com/Ericsson/codechecker/blob/master/web/tests/functional/comment/test_comment.py

You can use the `getReportDetails` API function to check whether it returns with the proper comments too. Similar to this:
https://github.com/Ericsson/codechecker/blob/47cf966edcafaaaf1047298d8fe0b03f0f3ca82a/web/tests/functional/extended_report_data/test_extended_report_data.py#L79

File: web/server/codechecker_server/api/report_server.py (hunk 1 above)
Comment (jay24rajput) [reply web/server/codechecker_server/api/report_server.py:636]: Yes, that totally makes sense. I am removing the file path line

File: web/server/codechecker_server/api/report_server.py (hunk 4 above)
Comment (jay24rajput) [reply web/server/codechecker_server/api/report_server.py]: I have done the changes. Kindly have a look. I am getting `%author%` and `%date` in the system comments. Is that okay??

File: web/server/codechecker_server/api/report_server.py [hunk 5]
```diff
@@ -623,11 +623,25 @@ def get_report_details(session, report_ids):
         extended_data.filePath = file_path
         extended_data_list[report_id].append(extended_data)
 
+    # Get Comments for report data
+    comment_data_list = defaultdict(list)
+    comment_query = session.query(Comment, Report.id, File.filepath)\
```
Comment (csordasmarton) [inline web/server/codechecker_server/api/report_server.py:628]: You are still querying the file path and joining the File table to this query which is unecessary. 
```suggestion
    comment_query = session.query(Comment, Report.id) \
```

File: web/server/codechecker_server/api/report_server.py [hunk 6]
```diff
@@ -623,11 +623,25 @@ def get_report_details(session, report_ids):
...
+        .filter(Report.id.in_(report_ids)) \
+        .outerjoin(Report, Report.bug_id == Comment.bug_hash) \
+        .outerjoin(File,
+                   File.id == Report.file_id)
+
+    for data, report_id, file_path in comment_query:
```
Comment (csordasmarton) [inline web/server/codechecker_server/api/report_server.py:634]:
```suggestion
    for data, report_id in comment_query:
```

File: web/server/codechecker_server/api/report_server.py [hunk 7]
```diff
@@ -623,11 +623,25 @@ def get_report_details(session, report_ids):
         extended_data.filePath = file_path
         extended_data_list[report_id].append(extended_data)
 
+    # Get Comments for report data
+    comment_data_list = defaultdict(list)
+    comment_query = session.query(Comment, Report.id, File.filepath)\
+        .filter(Report.id.in_(report_ids)) \
+        .outerjoin(Report, Report.bug_id == Comment.bug_hash) \
+        .outerjoin(File,
+                   File.id == Report.file_id)
```
Comment (csordasmarton) [inline web/server/codechecker_server/api/report_server.py:632]:
```suggestion
```

File: web/server/codechecker_server/api/report_server.py [hunk 8]
```diff
@@ -662,6 +676,34 @@ def extended_data_db_to_api(erd):
...
+    return ttypes.CommentData(
+        id=comm.id,
+        author=comm.author,
+        message=formatComments(comm),
+        createdAt=str(comm.created_at),
+        kind=comm.kind
+    )
+
+
+def formatComments(comment):
```
Comment (csordasmarton) [inline web/server/codechecker_server/api/report_server.py:689]: We are using snake_case function names in python code and I would also rename this function:
```suggestion
def get_comment_msg(comment):
```
Also please add a docstring to this function what this function does.

File: web/server/codechecker_server/api/report_server.py [hunk 9]
```diff
@@ -662,6 +676,34 @@ def extended_data_db_to_api(erd):
         fileId=erd.file_id)
 
 
+def comment_data_db_to_api(comm):
```
Comment (csordasmarton) [inline web/server/codechecker_server/api/report_server.py:679]: Please add a docstring what this function does.

**Review by csordasmarton — CHANGES_REQUESTED** [review]
I have also mentioned to add a new test case for this too.

**Comment by jay24rajput** [comment]
@csordasmarton I am still developing the tests for this functionality. Can you please guide me on which command should I use to run the test written in https://github.com/Ericsson/codechecker/blob/master/web/tests/functional/comment/test_comment.py

I have tried `make test`, `make test_functional` from the `~/codechecker` directory.
I have also tried `make test_functional` from `~/codechecker/web` directory, both the above commands some errors occur and the tests fail. Can you suggest some efficient and proper command to check the test written in the above file

**Comment by csordasmarton** [comment]
You can try to run the following command: `make TEST=tests/functional/comment test_web_feature`. It will run only the comment test cases and it will use sqlite database.

You can also find more information about our test cases here: https://github.com/Ericsson/codechecker/blob/master/docs/tests.md

**Comment by jay24rajput** [comment]
@csordasmarton I have added tests for the exporter. Kindly have a look and suggest changes if required

File: web/tests/functional/comment/test_comment.py [hunk 10]
```diff
@@ -275,3 +275,38 @@ def test_same_bug_hash(self):
         comments = self._cc_client.getComments(bug_new.reportId)
         user_comments, _ = separate_comments(comments)
         self.assertEqual(len(user_comments), 0)
+
+    def test_user_comments(self):
```
Comment (csordasmarton) [inline web/tests/functional/comment/test_comment.py:279]: This test case depends from the `test_comment` test above (this is the reason why there is a system comment in this test case), and if we rename the above mentioned function the test order may change. A better approach would be to move your checking after this line:
https://github.com/Ericsson/codechecker/blob/aefe6a68d35da56bc7e0a534d23aa53729a703b8/web/tests/functional/comment/test_comment.py#L183
We can do something similar to this and simplify the whole thing:
```py
details = self._cc_client.getReportDetails(bug.reportId)
user_c, system_c = separate_comments(details.comments)

self.assertEqual(len(user_c), len(user_comments))
self.assertEqual(user_c[0].message, user_comments[0].message)

self.assertEqual(len(system_c), len(system_comments))
self.assertEqual(system_c[0].message, system_comments[0].message)
```

**Review by csordasmarton — CHANGES_REQUESTED** [review]

File: web/tests/functional/comment/test_comment.py (hunk 10 above)
Comment (jay24rajput) [reply web/tests/functional/comment/test_comment.py]: Alright I have made the required changes 👍🏻

File: web/server/codechecker_server/api/report_server.py [hunk 11]
```diff
@@ -623,11 +623,23 @@ def get_report_details(session, report_ids):
         extended_data.filePath = file_path
         extended_data_list[report_id].append(extended_data)
 
+    # Get Comments for report data
+    comment_data_list = defaultdict(list)
+    comment_query = session.query(Comment, Report.id)\
+        .filter(Report.id.in_(report_ids)) \
+        .outerjoin(Report, Report.bug_id == Comment.bug_hash)
```
Comment (csordasmarton) [inline web/server/codechecker_server/api/report_server.py:630]: What do you think to sort the comment in an ascending order by the creation date? This way the first comment will always be the latest one.
```suggestion
        .outerjoin(Report, Report.bug_id == Comment.bug_hash) \
        .order_by(Comment.created_at.desc())
```

**Review by csordasmarton — COMMENTED** [review]
I tried it out and it works properly. I have just one tiny comment otherwise its LGTM!

File: web/server/codechecker_server/api/report_server.py (hunk 11 above)
Comment (jay24rajput) [reply web/server/codechecker_server/api/report_server.py]: There you go.. 😄

**Review by csordasmarton — APPROVED** [review]
LGTM! Thank you very much @jay24rajput your time and your work :blush:

**Pull request merged**